### ISSUE TITLE
JMS request/reply no possibility to reply with null

### ISSUE THREAD

**Issue by rroex** [spring-projects/spring-framework#23979]
When a JMSListener is used and a reply header is set, the return value is sent as a reply to the reply header destination. But if you return null, this behavior is ignored. According the documentation this is OK. But I am wondering if something could be added so that a reply will be sent even if return value is null. In my opinion, if the reply header is set, I expect a reply.

In [apache camel](https://camel.apache.org/components/latest/jms-component.html) they provide a "allowNullBody" option. Maybe a setting like this could help. 

I already tried to extend the MessagingMessageListenerAdapter and removed the if statement which checks if the return value is null. But then deeper in the code it fails because of Assert statements in GenericMessage and MessageBuilder. They expect always that the payload should not be null.

What do you think about this? @snicoll

**Comment by snicoll** [comment]
@rroex what does a `null` reply means exactly and why can't you return an empty String (i.e. empty `TextMessage`) or something like that?

**Comment by rroex** [comment]
For example you configure the jms listener as follow:

`@JmsListener(destination = "some.request")`
`public Object handle(Message<Object> message) {`
`return null;`
` }`

In this case, we expect a reply and the JMSReplyTo header is set.
But if we return null (because no result is found or whatever other reason) we now get a timeout.
This is because no reply is sent in case of a null result:

`if (result != null) {`
`handleResult(result, jmsMessage, session);`
`}`

See [this link](https://github.com/spring-projects/spring-framework/blob/master/spring-jms/src/main/java/org/springframework/jms/listener/adapter/MessagingMessageListenerAdapter.java#L78)

The expected output is a message with the content "null". If we would use Apache Camel instead, this would be the output.
To clarify, we use the JacksonMessageConverter.

**Comment by snicoll** [comment]
Thanks for the feedback, it looks like my question wasn't clear. Let me try again. Returning `null` is a signal to indicate that no reply should be sent and documented as you mentioned. I am not keen to change that and was asking why you could not return something that indicates an "empty message" for your use cases.

The current arrangement is better as you could indicate programmatically if a reply needs to be sent or not. Adding a flag that treats `null` as empty doesn't sound very appealing to me.

**Comment by rroex** [comment]
From my point of view, if the reply header is set, a reply message is expected. If `null` is returned, this means I expect a reply message with an empty payload. I think this would be the correct behaviour, because if you expect a reply, you are waiting for it. If it isn't been sent, you have to wait for a timeout that it failed. Looks strange to me. When you compare it to REST endpoints, you also get a reply if null is returned.

**Comment by snicoll** [comment]
@rroex sorry but that doesn't answer my question. As for your feedback on the fact you set a header, that's one way to look at it. The other way is that you put the onus on the framework to figure out what an empty message looks like. Depending on your serialization strategy that may mean different things that can be handled or not. Also, the idea of "null means no reply" is very easy to understand and ultimately the receiver of the original message should be in control of whether it sends a reply or not.

**Comment by evelknievel** [comment]
I have the same problem. I agree with @rroex that it should be possible to send an empty reply by returning null in the JmsListener. The client expects a reply. I dont want to write extra code to wrap the null result from my service etc.

**Comment by rroex** [comment]
It's like asking the shop if they still have stuff on stock, you still get a reply, even if they don't know.
Request/reply cycle is not completed in current implementation.

**Comment by evelknievel** [comment]
If you compare this for example with the rest endpoints, if you return null, still a reply is sent.

**Comment by rstoyanchev** [comment]
I don't think REST (request-reply) is a good analogy for JMS (asynchronous messaging).

**Comment by snicoll** [comment]
Indeed. Thanks for the suggestion but we're not going to pursue for several reasons:

* The server should be in control and specifying a header should not force in any way the server to send a reply. While we've been focusing on the corner case of handling `null` as an empty reply, we shouldn't forget about the feature of conditionally send a reply
* Conceptually, a message without a payload simply isn't a message in our model. The notion of what an empty message can't be determined at the framework level. What should `null` map to? An empty `TextMessage`? A byte array message with an empty `byte[]`?
* I've asked twice why returning an object that represents an empty message and didn't get an answer. Doing so addresses both points with an easy to understand programming model
